rustls/src/msgs/persist.rs

```diff
@@ -297,21 +297,205 @@ static MAX_FRESHNESS_SKEW_MS: u32 = 60 * 1000;
 
 // --- Server types ---
 #[derive(Debug)]
-pub struct ServerSessionValue {
-    pub(crate) sni: Option<DnsName<'static>>,
-    pub(crate) version: ProtocolVersion,
-    pub(crate) cipher_suite: CipherSuite,
+pub enum ServerSessionValue {
+    Tls12(Tls12ServerSessionValue),
+    Tls13(Tls13ServerSessionValue),
+}
+
+impl Codec<'_> for ServerSessionValue {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::Tls12(value) => {
+                ProtocolVersion::TLSv1_2.encode(bytes);
+                value.encode(bytes);
+            }
+            Self::Tls13(value) => {
+                ProtocolVersion::TLSv1_3.encode(bytes);
+                value.encode(bytes);
+            }
+        }
+    }
+
+    fn read(r: &mut Reader<'_>) -> Result<Self, InvalidMessage> {
+        match ProtocolVersion::read(r)? {
+            ProtocolVersion::TLSv1_2 => Ok(Self::Tls12(Tls12ServerSessionValue::read(r)?)),
+            ProtocolVersion::TLSv1_3 => Ok(Self::Tls13(Tls13ServerSessionValue::read(r)?)),
+            _ => Err(InvalidMessage::UnknownProtocolVersion),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Tls12ServerSessionValue {
+    #[doc(hidden)]
+    pub common: CommonServerSessionValue,
     pub(crate) master_secret: Zeroizing<PayloadU8>,
     pub(crate) extended_ms: bool,
+}
+
+impl Tls12ServerSessionValue {
+    pub(crate) fn new(
```

Comment:
Seems like `Tls12ServerSessionValue::new()` and `Tls13ServerSessionValue::new()` are pretty crappy as far as abstractions go, consider dropping them (assuming they don't have many callsites)?

Reply:
Yeah, that's what I was thinking.